lib/actions/app.ts

```diff
@@ -0,0 +1,227 @@
+limitations under the License.
+*/
+
+import { Application } from 'balena-sdk';
+import { CommandDefinition } from 'capitano';
+import * as commandOptions from './command-options';
+
+export const create: CommandDefinition<
+	{
+		name: string;
+	},
+	{
+		type?: string;
+	}
+> = {
+	signature: 'app create <name>',
+	description: 'create an application',
+	help: `\
+Use this command to create a new balena application.
+
+You can specify the application device type with the \`--type\` option.
+Otherwise, an interactive dropdown will be shown for you to select from.
+
+You can see a list of supported device types with
+
+	$ balena devices supported
+
+Examples:
+
+	$ balena app create MyApp
+	$ balena app create MyApp --type raspberry-pi\
+`,
+	options: [
+		{
+			signature: 'type',
+			parameter: 'type',
+			description:
+				'application device type (Check available types with `balena devices supported`)',
+			alias: 't',
+		},
+	],
+	permission: 'user',
+	async action(params, options, done) {
+		const balena = (await import('balena-sdk')).fromSharedOptions();
+
+		const patterns = await import('../utils/patterns');
+
+		// Validate the the application name is available
+		// before asking the device type.
+		// https://github.com/balena-io/balena-cli/issues/30
+		return balena.models.application
+			.has(params.name)
+			.then(hasApplication => {
+				if (hasApplication) {
+					return patterns.exitWithExpectedError(
+						'You already have an application with that name!',
+					);
+				}
+			})
+			.then(() => options.type || patterns.selectDeviceType())
+			.then(deviceType =>
+				balena.models.application.create({
+					name: params.name,
+					deviceType,
+				}),
+			)
+			.then(application =>
+				console.info(
+					`Application created: ${application.app_name} (${application.device_type}, id ${application.id})`,
+				),
+			)
+			.nodeify(done);
+	},
+};
+
+export const list: CommandDefinition = {
+	signature: 'apps',
+	description: 'list all applications',
+	help: `\
+Use this command to list all your applications.
+
+Notice this command only shows the most important bits of information for each app.
+If you want detailed information, use balena app <name> instead.
+
+Examples:
+
+	$ balena apps\
+`,
+	permission: 'user',
+	primary: true,
+	async action(_params, _options, done) {
+		const _ = await import('lodash');
+		const balena = (await import('balena-sdk')).fromSharedOptions();
+		const visuals = await import('resin-cli-visuals');
+
+		return balena.models.application
+			.getAll({
+				$select: ['id', 'app_name', 'device_type'],
+				$expand: { owns__device: { $select: 'is_online' } },
+			})
+			.then(
+				(
+					applications: Array<
+						Application & { device_count?: number; online_devices?: number }
+					>,
+				) => {
+					applications.forEach(application => {
+						application.device_count = _.size(application.owns__device);
+						application.online_devices = _.sumBy(application.owns__device, d =>
```

Comment:
I switched this from a `.filter(...).length` to this `.sumBy(...)` to avoid creating an intermediary array unnecessarily, it probably doesn't matter for the CLI specifically but I have seen real performance hits from it so as a best practice I converted it